src/x86_64/mp.c

```diff
@@ -0,0 +1,78 @@
+//#define MP_DEBUG
+
+#include <x86_64.h>
+#include <apic.h>
+#include <page.h>
+
+static void *apboot = 0;
+extern u8 apinit, apinit_end;
+extern void *ap_pagetable, *ap_start_vector, *ap_gdt_pointer, *ap_idt_pointer, *ap_stack;
+extern u64 ap_lock;
+
+#define ICR_TYPE_INIT         0x00000500
+#define ICR_TYPE_STARTUP      0x00000600
+
+static heap pages;
+static void (*start_callback)();
+
+#ifdef MP_DEBUG
+#define mp_debug(x) console(x)
+#define mp_debug_u64(x) print_u64(x)
+#else
+#define mp_debug(x)
+#define mp_debug_u64(x)
+#endif
+
+static void __attribute__((noinline)) ap_new_stack()
+{
+    mp_debug("ap_new_stack for cpu ");
+
+    u64 id = apic_id();
+    mp_debug_u64(id);
+    cpu_setgs(id);
+    cpuinfo ci = current_cpu();
+
+    set_ist(id, IST_PAGEFAULT, u64_from_pointer(ci->fault_stack));
+    set_ist(id, IST_INTERRUPT, u64_from_pointer(ci->int_stack));
+    set_running_frame(ci->misc_frame);
+    mp_debug(", install gdt");
+    install_gdt64_and_tss(id);
+    mp_debug(", enable apic");
+    enable_apic();
+    mp_debug(", clear ap lock, enable ints, start_callback\n");
+    ap_lock = 0;
+    enable_interrupts();
+    start_callback();
+}
+
+void ap_start()
+{
+    void *n = allocate_stack(pages, 16);
+    asm volatile("mov %0, %%rsp": :"m"(n));
+    ap_new_stack();
+}
+
+void start_cpu(heap h, heap p, int index, void (*ap_entry)()) {
+    pages = p ;
+    start_callback = ap_entry;
+    if (!apboot) {
+        apboot =  (void *)0x8000;
+        map((u64)apboot, (u64)apboot, 4096, PAGE_WRITABLE, h);
+    }
+    asm("sgdt %0": "=m"(ap_gdt_pointer));
+    asm("sidt %0": "=m"(ap_idt_pointer));    
+    mov_from_cr("cr3", ap_pagetable);
+    // just one function call
```

Comment:
just going to mention that there is plenty of room at 0x8000 to make this call, so we don't really need to leak this page..but it doesn't really .. matter?

Reply:
It kind of is an issue since that area is supposed to be open to userspace...right now there's nothing to prevent a collision with it.

So maybe we can set some deadline for the APs to come up and then just unmap it?

Also - any reason that ap.s is located in boot when it's linked into stage3?

Reply:
well, we should fix it. ideally by using an allocator of some kind. I can look it up again, but I think it needs to be located in the lowest 1<<20? 

no, absolutely no reason...it wasn't clear how it was going to land when I started and it seemed to be more related to boot than anything else. and I was quite surprised that the makefile just worked..x86_86 is fine

Reply:
what about 0? that's virtually unusable elsewhere, and would solve the offset issue :)

Reply:
So I was able to keep the address defined in a single place - the linker script. Which is kind of odd, because nothing actually gets linked at that address, but it does allow us to have a single place to define it.

Surprisingly, the zero page works. I say surprisingly because I thought the real mode interrupt vectors were in that page. But maybe it doesn't matter for the whole ap trampoline scheme? In any case, it's easy enough to move around now.